### DISCUSSION TITLE
FMU leads to MATLAB/Simulink crash on every consecutive simulation

### DISCUSSION THREAD

**Discussion by AishwaryaRaghavendra** [StephenSmith25/PythonFMU3#1] [Q&A]
I have used pythonfmu3 to build FMU from my python script including data science packages in windows. Command used to successfully build FMU: pythonfmu3 build -f python_script.py requirements.txt

Further I had also used fmpy package to simulate this FMU and I also get the results for the given inputs as I expected.

My end goal is to import this FMU in Matlab/Simulink and simulate. I was again successfull in importing FMU and executing in Simulink for the first simulation run. But in the consecutive (second) simulation run, the Simulink/Matlab crashed for unknown issue. Then I have to restart my Matlab/Simulink and simulate again where it works for the first time but crashes again in the second time. So I keep restarting my Matlab/Simulink everytime to make FMU work.

FYI, there is no problem with Matlab version and compatibility with FMU. Othe simulations work without any problem. Also this FMU works for the first simulation run after restart.

Generally, does the FMU instance requires to be reset/free/terminate like I did when using fmpy package even within Matlab after every simulation run?

Do you know what the problem can be with FMU that works using fmpy package but not in Matlab everytime ?

**Comment by StephenSmith25** [comment]
Sorry, I wonder if there is a bug with reset/free. I did this project as a quick MVP to implement FMI3 support for PythonFmu. I will have to check. 

I unfortunately have not used MATLAB in years. If you have any log messages or debug information you could share with me that would be great.

Thanks.

**Reply by axellehego** [reply]
Hello,
I have exactly the same problem as [AishwaryaRaghavendra](https://github.com/AishwaryaRaghavendra). I tried to run a second time FMUs generated with pythonfmu3 (0.1.11) on Simulink R2023b which support FMU3.0 and Simulink just close.
I tried also to simulate the same FMUs with Dymola 2023x Refresh 1 which support also FMU3.0 and I have a similar problem when I try to recompile the FMU. On Dymola, we have an .exe generated at each compilation of the model which is exectued in simulation/run stage. The .exe normally closes when the simulation is finished but it isn't the case with pythonfmu3's FMUs. This causes the following error "Failed to delete dymosim.exe […] Complilation failed."
I don't know if this can help you to find the problem.

**Reply by StephenSmith25** [reply]
Just to clarify, does it work on the first run? There are some discussions in the parent project that relate to using external tools with a python FMU. The challenge is the FMU needs python to run and specifically it needs libpython. I wonder if the comments at the end of github.com/NTNU-IHB/PythonFMU/issues/208 helps.

I can try debugging things if you have a more verbose log. But a few things you could try
1) does the FMU work with fmpy
2) can you run an FMU generated with pythonfmu (so version 2 fmus)
3) does it work with a really simple FMU generated by pythonfmu3 (i.e. one with no dependencies on other libraries, e.g numpy)

Unfortunately I cannot test this as I do not have access to either of the tools you mentioned.

**Reply by axellehego** [reply]
Yes the first run as expected but when I try to run again Simulink crash (Dymola > Recompile > Error).
1. I didn't try fmpy because I will use this FMU with Dymola and Simulink
2. Yes, FMU2.0 generated with pythonfmu run the first time and with recompilation without this problems with Dymola but crash at the second run on Simulink
3. No, I tried one of the example. I have download bouncingBall.py, used pythonfmu3 build -f bouncingBall.py in a python environment  where I installed pythonfmu3==0.1.11. Than I load the FMU in Dymola and Simulate (see attached .pdf and the log file dslog.txt)

[Test_FMU_pythonfmu3.pdf](https://github.com/StephenSmith25/PythonFMU3/files/15013361/Test_FMU_pythonfmu3.pdf)
[dslog.txt](https://github.com/StephenSmith25/PythonFMU3/files/15013486/dslog.txt)
